File: 03.Sliding Window/121. Best Time to Buy and Sell Stock.cpp
// Solution1 (NeetCode) faster version
/*
    Given array prices, return max profit w/ 1 buy & 1 sell
    Ex. prices = [7,1,5,3,6,4] -> 5 (buy at $1, sell at $6)

    For each, get diff b/w that & min value before, store max

    Time: O(n)
    Space: O(1)
*/
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minValue = prices[0];
        int maxDiff = 0;
        
        for (int i = 1; i < prices.size(); i++) {
            minValue = min(minValue, prices[i]);
            maxDiff = max(maxDiff, prices[i] - minValue);
        }
        
        return maxDiff;
    }
};



// Solution2 (My implementation)
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int res=0;
        int buy=0, sell=1;
        
        while (sell < prices.size()) {
            res = max(res, prices[sell]-prices[buy]);
            if (prices[sell] < prices[buy]) {
                buy = sell;
            }
            sell++;
        }
        return res;
    }
};

File: 03.Sliding Window/121. Best Time to Buy and Sell Stock.py
### Solution1 (NeetCode faster) ###
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0

        l = 0
        for r in range(1, len(prices)):
            if prices[r] < prices[l]:
                l = r
            res = max(res, prices[r] - prices[l])
        return res



### Solution2 (My method) ###
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy, sell = 0,1
        L = len(prices)
        res = 0
        
        # while buy<L and sell<L and buy<sell:
        while sell<L:
            profit = prices[sell]-prices[buy]
            res = max(res, profit)
            if prices[sell] < prices[buy]:
                buy=sell
            sell+=1
        return res